[libSyntax] Create deferred nodes in SyntaxParseActions

ahoppen opened the pull request:
Move the creation of deferred nodes from the `ParsedRawSyntaxRecorder` to `SyntaxParseActions`. By itself this does not offer any advantage, but from here we will be able to optimize the creation of deferred nodes inside the different `SyntaxParseActions` implementations.

All implementation that is currently in `SyntaxParseActions` is transitional. In the next PR these will be pushed to `CLibParseActions` (which will mostly match what’s currently in `SyntaxParseActions`) and `SyntaxTreeCreator` (which can avoid creating deferred nodes altogether by directly constructing `RawSyntax` nodes).

ahoppen committed:
[libSyntax] Create DeferredLayoutNode/DeferredTokenNode in ParsedRawSyntaxRecorder

This is an intermediate step in the effort to push deferred node
handling all the way down to SyntaxParseActions.

ahoppen committed:
[libSyntax] Remove unnecessary getOpaqueNode and takeOpaqueNode from ParsedRawSyntaxNode

They are superseded by getData/takeData

ahoppen committed:
[libSyntax] Re-enable dumping of ParsedRawSyntaxNode children

This requires a SyntaxParsingContext to be passed to the dump function
and was thus disabled in an earlier commit

ahoppen commented:
@swift-ci Please smoke test

ahoppen commented:
apple/swift-syntax#260

@swift-ci Please smoke test

ahoppen commented:
apple/swift-syntax#260

@swift-ci Please smoke test

ahoppen commented:
apple/swift-syntax#260

@swift-ci Please smoke test

ahoppen commented:
apple/swift-syntax#260

@swift-ci Please smoke test Linux

ahoppen commented:
@swift-ci Please smoke test

ahoppen commented:
apple/swift-syntax#260

@swift-ci Please smoke macOS

ahoppen commented:
apple/swift-syntax#260

@swift-ci Please smoke test macOS

ahoppen commented:
apple/swift-syntax#260

@swift-ci Please smoke test

ahoppen commented:
@swift-ci Please smoke test

rintaro commented on include/swift/Parse/ParsedRawSyntaxNode.h:
```diff
@@ -150,6 +102,24 @@ class ParsedRawSyntaxNode {
+  /// null).
+  DataKind getNodeType() const { return Data.getKind(); }
+
+  /// Returns the opaque data of this node. This must be interpreted by the
+  /// \c SyntaxParseAction, which likely also needs the node type to know
+  /// what type of node the data represents.
+  OpaqueSyntaxNode getData() const { return Data.getOpaque(); }
+
+  RecordedOrDeferredNode getRecordedOrDeferredNode() const { return Data; }
```
These methods (`getData()` and `getRecordedOrDeferredNode()`) breaks the "move-only" semantics of `ParsedRawSyntaxNode`. "Data" should not be extracted without resetting the `ParsedRawSyntaxNode`.
This is important to ensure not to discard parsed recorded nodes without putting them into the final tree.

Previously, we allowed copying via `copyDeferred()` method only if it's deferred.

ahoppen replied:
We need some kind of “get data, but don’t take” method to retrieve the node’s children [here](https://github.com/apple/swift/pull/36277/files#diff-a83cf4327ae0cebc7c22c773cfa0aac6304f414828248800d7492f1e3c4145a8R214).

I’ve renamed the method to `getUnsafeOpaqueData` and added a comment that the data returned by it should only be used to inspect the node, not to record it. Are you OK with that?

Also, while looking through the uses of this method, I actually found a case where we should have been using `takeData` instead, so good catch.

rintaro commented on lib/Parse/SyntaxParseActions.cpp:
```diff
@@ -0,0 +1,170 @@
+    case RecordedOrDeferredNode::Kind::Null:
+      break;
+    case RecordedOrDeferredNode::Kind::Recorded:
+      llvm_unreachable("Children of deferred nodes must also be deferred");
```
Do we have any comments why deferred nodes can't contain recorded nodes? If not, could you add some?

ahoppen replied:
It’s because of the way we create the syntax tree. Once we put the `SyntaxParsingContext` in deferred node all (child) nodes are also deferred. If we chose to record child nodes of deferred nodes, we would also get into trouble should we decide to discard the deferred node, because now we’d also need to discard the recorded child node, which cannot be done. Please correct me if any of the assumptions I’m making here are wrong.

We don’t have a comment about this anywhere. I’ll add one with (more or less) the text I mentioned above.

rintaro commented on lib/Parse/SyntaxParseActions.cpp:
```diff
@@ -0,0 +1,170 @@
+    auto Child = Data->Children[i];
+    switch (Child.getKind()) {
+    case RecordedOrDeferredNode::Kind::Null:
+    case RecordedOrDeferredNode::Kind::Recorded:
```
Other methods states "Children of deferred nodes must also be deferred". So this should be unreachable, right?

ahoppen replied:
Yes, you’re correct.

ahoppen commented on include/swift/Parse/ParsedRawSyntaxNode.h:
```diff
@@ -150,6 +102,24 @@ class ParsedRawSyntaxNode {
+
+  RecordedOrDeferredNode getRecordedOrDeferredNode() const { return Data; }
+
+  /// Return the opaque data of this node and reset it.
+  OpaqueSyntaxNode takeData() {
+    OpaqueSyntaxNode Data = this->getData();
+    reset();
+    return Data;
+  }
```
None, `takeOpaqueNode` will be removed.

ahoppen committed:
[libSyntax] Remove getData method from ParsedRawSyntaxNode

The getData and getRecordedOrDeferredNode methods broke the move-only semantics of ParsedRawSyntaxNode because it allowed retrieving the data without resetting it.

Change most uses to use takeData or takeRecordedOrDeferredNode instead of the get methods.

To retrieve the children of a ParsedRawSyntaxNode, we still need a way to access the OpaqueSyntaxNode for inspection by the SyntaxParseActions without resetting it. For this, we still maintain a method with the semantics of getData, but it’s now called getUnsafeOpaqueData to make sure it’s not accidentally used.

ahoppen commented:
@swift-ci Please smoke test

ahoppen committed:
[libSyntax] Handle deferred node data in SyntaxParseActions

By now ParsedRawSyntaxNode does not have any knowledge about deferred
node data anymore, which frees up SyntaxParseActions (and, in
particular its sublass SyntaxTreeCreator) to perform optimisations to
more efficiently create and record deferred nodes.

ahoppen committed:
[libSyntax] Store DataKind and OpaqueSyntaxNode in ParsedRawSyntaxNode as RecordedOrDeferredNode

This is mostly a clean-up commit. ParsedRawSyntaxNode stores an
OpaqueSyntaxNode and its DataKind so it might as well use a
RecordedOrDeferredNode.

ahoppen commented:
@swift-ci Please smoke test

rintaro commented on lib/Parse/ParsedSyntaxNodes.cpp.gyb:
```diff
@@ -27,15 +27,15 @@ using namespace swift::syntax;
-Parsed${child.type_name} Parsed${node.name}::getDeferred${child.name}() {
-  return Parsed${child.type_name} {getRaw().getDeferredChildren()[${node.name}::Cursor::${child.name}].copyDeferred()};
+Parsed${child.type_name} Parsed${node.name}::getDeferred${child.name}(const SyntaxParsingContext *SyntaxContext) {
+  return Parsed${child.type_name} {getRaw().getDeferredChild(${node.name}::Cursor::${child.name}, SyntaxContext).copyDeferred()};
```
`copyDeferred()` is not needed I believe

ahoppen replied:
It’s still used in one place, I’ll remove it in https://github.com/apple/swift/pull/36362/commits/81f7d1e0d7ef81cbcb21a9b757eb664e27bb42a9 which is part of https://github.com/apple/swift/pull/36362

rintaro commented on lib/Parse/ParsedSyntaxNodes.cpp.gyb:
```diff
@@ -27,15 +27,15 @@ using namespace swift::syntax;
+  auto RawChild = getRaw().getDeferredChild(${node.name}::Cursor::${child.name}, SyntaxContext);
   if (RawChild.isNull())
     return None;
   return Parsed${child.type_name} {RawChild.copyDeferred()};
```
ditto

rintaro replied:
If you remove them, `copyDeferred()` is not used anymore...

ahoppen replied:
See above.

ahoppen committed:
[libSyntax] Allow retrieval of a null child from deferred ParsedRawSyntaxNodes

ahoppen commented:
@swift-ci Please smoke test

ahoppen commented:
@swift-ci Please smoke test Windows

ahoppen commented:
@swift-ci Please smoke test Windows
